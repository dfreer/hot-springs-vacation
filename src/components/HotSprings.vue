<template>
  <section>
    <h2 class="text-2xl m-auto mb-8 max-w-5/6">
      Please enter your desired 2-D grid size:
    </h2>
    <form @submit.prevent="createGrid">
      <label for="height" class="text-lg block">Height:</label>
      <input
        v-model.number="heightInput"
        id="height"
        type="number"
        class="text-lg border-2 rounded my-2 p-1"
        :disabled="springsInProgress"
      />
      <label for="width" class="text-lg block">Width:</label>
      <input
        v-model.number="widthInput"
        id="width"
        type="number"
        class="text-lg border-2 rounded my-2 p-1"
        :disabled="springsInProgress"
      />
      <button class="button-lg mt-4" :disabled="springsInProgress">Create Grid</button>
      <p
        :class="[
          error ? 'opacity-1' : 'opacity-0 hidden',
          'transition-all',
          'duration-1000',
          'ease-in-out',
          'bg-red-200',
          'mx-auto',
          'my-6',
          'max-w-3/4',
          'md:max-w-1/2',
          'md:h-20',
          'h-16',
          'flex',
          'justify-center',
          'items-center',
          'rounded-lg',
        ]"
      >
        {{ error }}
      </p>
      <Grid v-if="isGrid" />
    </form>
  </section>
</template>

<script>
import { mapGetters } from 'vuex'
import Grid from './Grid.vue'

export default {
  name: 'HotSprings',
  components: { Grid },
  data: () => ({
    heightInput: null,
    widthInput: null,
  }),
  computed: { ...mapGetters(['springsInProgress', 'isGrid', 'error']) },
  methods: {
    createGrid() {
      if (!Number.isInteger(this.heightInput) || !Number.isInteger(this.widthInput)) {
        this.$store.dispatch('showError', 'Please provide width and height')
        return
      }

      if (this.heightInput > 20 || this.widthInput > 20) {
        this.$store.dispatch(
          'showError',
          'Number provided is too big, please provide a number less than 20',
        )
        this.heightInput = null
        this.widthInput = null
        return
      }

      if (this.heightInput < 1 || this.widthInput < 1) {
        this.$store.dispatch(
          'showError',
          'Number provided is too small, please provide a number greater than 1',
        )
        this.heightInput = null
        this.widthInput = null
        return
      }

      this.$store.commit('SET_SPRINGS_DATE', null)
      this.$store.commit('SET_FULL_SPRINGS_REACH', false)
      this.$store.commit('SET_ERROR', null)
      this.$store.dispatch('createGrid', {
        height: this.heightInput,
        width: this.widthInput,
      })

      this.heightInput = null
      this.widthInput = null
    },
  },
}
</script>
