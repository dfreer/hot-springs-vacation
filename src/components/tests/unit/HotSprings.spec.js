import { shallowMount } from '@vue/test-utils'
import HotSprings from '../../HotSprings.vue'

// defining required properties to mock them later
const requiredComputed = {
  springsInProgress() {
    return false
  },
  error() {
    return false
  },
  isGrid() {
    return false
  }
}

const storeMock = {
  getters: {}
}

// 🕵️
const storeMockWithDispatch = {
  ...storeMock,
  dispatch: jest.fn()
}

// 🕵️🕵️
const storeMockWithDispatchAndCommit = {
  ...storeMockWithDispatch,
  commit: jest.fn()
}

describe('HotSprings', () => {
  it('mounts successfully', () => {
    const wrapper = shallowMount(HotSprings, {
      mocks: { $store: storeMock },
      computed: requiredComputed
    })
    expect(wrapper.exists()).toBeTruthy()
  })
  it('matches snapshot', () => {
    const wrapper = shallowMount(HotSprings, {
      mocks: { $store: storeMock },
      computed: requiredComputed
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('HotSprings error handler', () => {
  // displays error if height and width are not integers
  it('displays error if height and width are floats', () => {
    const wrapper = shallowMount(HotSprings, {
      mocks: { $store: storeMockWithDispatch },
      data: () => ({
        heightInput: 0.5,
        widthInput: 0.5
      }),
      computed: requiredComputed
    })
    wrapper.vm.createGrid()
    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith(
      'showError',
      'Please provide width and height'
    )
  })
  it('displays error if height and width are > 20', () => {
    const wrapper = shallowMount(HotSprings, {
      mocks: { $store: storeMockWithDispatch },
      data: () => ({
        heightInput: 25,
        widthInput: 25
      }),
      computed: requiredComputed
    })
    wrapper.vm.createGrid()
    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith(
      'showError',
      'Number provided is too big, please provide a number less than 20'
    )
  })
  it('displays error if height and width are < 0', () => {
    const wrapper = shallowMount(HotSprings, {
      mocks: { $store: storeMockWithDispatch },
      data: () => ({
        heightInput: -5,
        widthInput: -5
      }),
      computed: requiredComputed
    })
    wrapper.vm.createGrid()
    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith(
      'showError',
      'Number provided is too small, please provide a number greater than 1'
    )
  })
  it('displays error if height and width are 0', () => {
    const wrapper = shallowMount(HotSprings, {
      mocks: { $store: storeMockWithDispatch },
      data: () => ({
        heightInput: 0,
        widthInput: 0
      }),
      computed: requiredComputed
    })
    wrapper.vm.createGrid()
    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith(
      'showError',
      'Number provided is too small, please provide a number greater than 1'
    )
  })
  it('invokes proper commits with proper height and width input', () => {
    const wrapper = shallowMount(HotSprings, {
      mocks: { $store: storeMockWithDispatchAndCommit },
      data: () => ({
        heightInput: 5,
        widthInput: 5
      }),
      computed: requiredComputed
    })
    wrapper.vm.createGrid()

    expect(wrapper.vm.$store.commit).toHaveBeenNthCalledWith(1, 'SET_SPRINGS_DATE', null)
    expect(wrapper.vm.$store.commit).toHaveBeenNthCalledWith(
      2,
      'SET_FULL_SPRINGS_REACH',
      false
    )
    expect(wrapper.vm.$store.commit).toHaveBeenNthCalledWith(3, 'SET_ERROR', null)

    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith('createGrid', {
      height: 5,
      width: 5
    })
  })
})
