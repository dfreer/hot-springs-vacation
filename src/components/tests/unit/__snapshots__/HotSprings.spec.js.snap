// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotSprings matches snapshot 1`] = `
<section>
  <h2
    class="text-2xl m-auto mb-8 max-w-5/6"
  >
    
    Please enter your desired 2-D grid size:
  
  </h2>
   
  <form>
    <label
      class="text-lg block"
      for="height"
    >
      Height:
    </label>
     
    <input
      class="text-lg border-2 rounded my-2 p-1"
      id="height"
      type="number"
    />
     
    <label
      class="text-lg block"
      for="width"
    >
      Width:
    </label>
     
    <input
      class="text-lg border-2 rounded my-2 p-1"
      id="width"
      type="number"
    />
     
    <button
      class="button-lg mt-4"
    >
      Create Grid
    </button>
     
    <p
      class="opacity-0 hidden transition-all duration-1000 ease-in-out bg-red-200 mx-auto my-6 max-w-3/4 md:max-w-1/2 md:h-20 h-16 flex justify-center items-center rounded-lg"
    >
      
      false
    
    </p>
     
    <!---->
  </form>
</section>
`;
